File: app/src/main/java/net/nashihara/naroureader/presenter/SearchPresenter.java
package net.nashihara.naroureader.presenter;

import android.support.annotation.Nullable;

import net.nashihara.naroureader.entities.Query;
import net.nashihara.naroureader.views.SearchView;

import java.util.ArrayList;

import narou4j.enums.NovelGenre;

public class SearchPresenter implements Presenter<SearchView> {

    private SearchView view;

    public SearchPresenter(SearchView view) {
        attach(view);
    }

    @Override
    public void attach(SearchView view) {
        this.view = view;
    }

    @Override
    public void detach() {
        view = null;
    }

    private ArrayList<Integer> validateGenreIds(boolean[] genreChecked) {
        NovelGenre[] genreIds = NovelGenre.values();
        ArrayList<Integer> genres = new ArrayList<>();
        for (int i = 0; i < genreChecked.length; i++) {
            if (genreChecked[i]) {
                genres.add(genreIds[i].getId());
            }
        }

        return genres;
    }

    public void shapeSearchQuery(@Nullable Query query, @Nullable boolean[] genreChecked) {
        if ((query != null ? query.getLimit() : 0) > 500) {
            view.showError();
            return;
        }

        ArrayList<Integer> genres = validateGenreIds(genreChecked);

        view.showResult(query , genres);
    }
}

File: app/src/main/java/net/nashihara/naroureader/presenter/SearchPresenter.kt
package net.nashihara.naroureader.presenter

import net.nashihara.naroureader.entities.Query
import net.nashihara.naroureader.views.SearchView

import java.util.ArrayList

import narou4j.enums.NovelGenre

class SearchPresenter(view: SearchView) : Presenter<SearchView> {

    private var view: SearchView? = null

    init {
        attach(view)
    }

    override fun attach(view: SearchView) {
        this.view = view
    }

    override fun detach() {
        view = null
    }

    private fun validateGenreIds(genreChecked: BooleanArray?): ArrayList<Int> {
        val genreIds = NovelGenre.values()
        val genres = ArrayList<Int>()

        genreChecked?.indices
                ?.filter { genreChecked[it] }
                ?.mapTo(genres) { genreIds[it].id }

        return genres
    }

    fun shapeSearchQuery(query: Query?, genreChecked: BooleanArray?) {
        if (query?.limit ?: 0 > 500) {
            view?.showError()
            return
        }

        val genres = validateGenreIds(genreChecked)

        view?.showResult(query!!, genres)
    }
}
